misk-crypto/src/main/kotlin/misk/crypto/FieldLevelEncryptionPacket.kt

```diff
@@ -0,0 +1,358 @@
+   */
+  enum class ContextKey constructor(val index: Int) {
+    UNDEFINED(1 shl 0),
+    TABLE_NAME(1 shl 1),
+    DATABASE_NAME(1 shl 2),
+    COLUMN_NAME(1 shl 3),
+    SHARD_NAME(1 shl 4),
+    PRIMARY_ID(1 shl 5),
+    EVENT_TOPIC(1 shl 6),
+    SERVICE_NAME(1 shl 7),
+    CUSTOMER_TOKEN(1 shl 8),
+  }
+
+  /**
+   * Types of possible TLV packets
+   */
+  enum class EntryType(val type: Int) {
+    UNDEFINED(0),
+
+    /**
+     * A description of a context that can have missing values (but all keys that are included). Also leaves out fields
+     * marked in the bitmask.
+     */
+    EXPANDED_CONTEXT_DESCRIPTION(1),
+
+    /**
+     * The complete serialized AAD, what is required for AEAD.
+     */
+    COMPLETE_AAD(2),
+
+    /**
+     * A payload that has a size field; used for storing packets that we'd like to append to.
+     */
+    PAYLOAD_SIZED(3),
+
+    /**
+     * A marker for a payload that just goes until end of packet.
+     */
+    PAYLOAD_REMAINDER(4)
+  }
+
+  /**
+   * Possible inner fields a packet can have.
+   *
+   * TODO(yivnitskiy): Move 'parsing' logic from the 'when' below to each type
+   */
+  sealed class Field(open val payload: ByteArray) {
+    data class ExpandedContextDescription(override val payload: ByteArray) : Field(payload) {
+      companion object {
+        val type: EntryType = EntryType.EXPANDED_CONTEXT_DESCRIPTION
+      }
+
+      override fun type() = type
+    }
+
+    data class CompleteAad(override val payload: ByteArray) : Field(payload) {
+      companion object {
+        val type: EntryType = EntryType.COMPLETE_AAD
+      }
+
+      override fun type() = type
+    }
+
+    data class PayloadSized(override val payload: ByteArray) : Field(payload) {
+      companion object {
+        val type: EntryType = EntryType.PAYLOAD_SIZED
+      }
+
+      override fun type() = type
+    }
+
+    data class PayloadRemainder(override val payload: ByteArray) : Field(payload) {
+      companion object {
+        val type: EntryType = EntryType.PAYLOAD_REMAINDER
+      }
+
+      override fun type() = type
+
+      override fun writeToStreamIfNotEmpty(s: DataOutput) {
+        if (payload.isEmpty())
+          return
+
+        s.writeByte(this.type().type)
+        s.write(payload)
+      }
+    }
+
+    abstract fun type(): EntryType
+
+    open fun writeToStreamIfNotEmpty(s: DataOutput) {
+      if (payload.isEmpty())
+        return
+
+      if (payload.size > Short.MAX_VALUE) {
+        throw InvalidEncryptionContextException("Field is too large (${payload.size} is greater than ${Short.MAX_VALUE})")
+      }
+
+      s.writeByte(this.type().type)
+      s.writeShort(payload.size)
+      s.write(payload)
+    }
+
+    companion object {
+      fun readField(s: DataInputStream): Field? {
+        val type = s.read()
+        if (type < 0)
+          return null
+        return when (type) {
+          ExpandedContextDescription.type.type -> {
+            val size = s.readUnsignedShort()
```

Comment:
annoyingly, readUnsignedShort is `readShort() & 0xffff` if you’re using BufferedSink. Still the right approach though I think